Remove legacy jerry_get_memory_limits API function and unused configuration macros

akosthekiss opened the pull request:
Also remove the legacy `jerry_get_memory_limits` API function as it
does not report either useful or valid information.

JerryScript-DCO-1.0-Signed-off-by: Akos Kiss akiss@inf.u-szeged.hu

akosthekiss commented:
**NOTE: THIS IS AN API-BREAKING CHANGE!** (At least, theoretically.)

In practice, I cannot think of any user of JerryScript who would have used the `jerry_get_memory_limits` function. It has been introduced in Feb 11, 2015 (35fa39ceb8fc79b6ce33c07d753f3ff2c8d3d39a) but hasn't been of any use since Feb 13, 2015 (43ea53b1d7aad510bc547fc30c9912daddb6828b), when its counterpart `jrt_set_mem_limits` was macrod out to never get enabled ever again.

akosthekiss commented:
Speaking of config macros and breaking changes: is the `CONFIG_ECMA_GLOBAL_ENVIRONMENT_DECLARATIVE` setting used / useful? The only place where this is used at is the following. Which doesn't seem to be too standard. It's also been introduced several years ago (Sep 2014 (fbcd393962decd30fb2eb51f92e63dfa92031046)) but I'm not aware of any (current?) use cases.

```c
void
ecma_init_global_lex_env (void)
{
#ifdef CONFIG_ECMA_GLOBAL_ENVIRONMENT_DECLARATIVE
  JERRY_CONTEXT (ecma_global_lex_env_p) = ecma_create_decl_lex_env (NULL);
#else /* !CONFIG_ECMA_GLOBAL_ENVIRONMENT_DECLARATIVE */
  ecma_object_t *glob_obj_p = ecma_builtin_get (ECMA_BUILTIN_ID_GLOBAL);

  JERRY_CONTEXT (ecma_global_lex_env_p) = ecma_create_object_lex_env (NULL, glob_obj_p, false);

  ecma_deref_object (glob_obj_p);
#endif /* CONFIG_ECMA_GLOBAL_ENVIRONMENT_DECLARATIVE */
} /* ecma_init_global_lex_env */
```

I'm happy to include the removal of the macro and the associated code paths in this patch but feedback would be appreciated on this first.

zherczeg commented:
#2213

zherczeg commented:
I would emphasize the removal of `jerry_get_memory_limits` rather than macros in the title and commit log. Other than that is ok for me.

akosthekiss commented:
@zherczeg And what about the question raised in the comment: what's the deal with CONFIG_ECMA_GLOBAL_ENVIRONMENT_DECLARATIVE ?

zherczeg commented:
That macro makes an ecmascript program behave like a function without the argument and `return` support. It is non-standard.

Currently you can create functions directly, and I would recommend that if you want to use such a feature.

akosthekiss commented:
I've updated the PR according to the feedbacks:
- PR title and commit message changed
- removed CONFIG_ECMA_GLOBAL_ENVIRONMENT_DECLARATIVE from the code base
- rebased onto latest master
